crates/accelerate/src/synthesis/cnot_synthesis.rs

```diff
@@ -0,0 +1,163 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2024
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use hashbrown::HashMap;
+use ndarray::{s, Array1, Array2, Axis};
+use numpy::PyReadonlyArray2;
+use smallvec::smallvec;
+use std::cmp;
+
+use qiskit_circuit::circuit_data::CircuitData;
+use qiskit_circuit::operations::{Param, StandardGate};
+use qiskit_circuit::Qubit;
+
+use pyo3::prelude::*;
+
+/// Mutate a matrix inplace by adding the value of the ``ctrl`` row to the
+/// ``target`` row. If ``add_columns`` is true, add columns instead of rows.
+fn _add(mat: &mut Array2<bool>, add_columns: &bool, ctrl: usize, trgt: usize) {
+    // get the two rows (or columns)
+    let info = if *add_columns {
+        (s![.., ctrl], s![.., trgt])
+    } else {
+        (s![ctrl, ..], s![trgt, ..])
+    };
+    let (row0, mut row1) = mat.multi_slice_mut(info);
+
+    // add them inplace
+    row1.zip_mut_with(&row0, |x, &y| *x ^= y);
+}
+
+/// This helper function allows transposed access to a matrix.
+fn _index(transpose: &bool, i: &usize, j: &usize) -> (usize, usize) {
+    if *transpose {
+        (*j, *i)
+    } else {
+        (*i, *j)
+    }
+}
+
+/// Synthesize a linear function, given by a boolean square matrix, into a circuit.
+/// This function uses the Patel-Markov-Hayes algorithm, described in arXiv:quant-ph/0302002,
+/// using section-wise elimination of the rows.
+#[pyfunction]
+#[pyo3(signature = (matrix, section_size=2))]
+pub fn synth_cnot_count_full_pmh(
+    py: Python,
+    matrix: PyReadonlyArray2<bool>,
+    section_size: i64,
+) -> PyResult<CircuitData> {
+    let mut mat: Array2<bool> = matrix.as_array().to_owned();
+    // let mut mat = matrix.to_owned();
+    let num_qubits = mat.nrows(); // is a quadratic matrix
+    let lower_cnots = lower_cnot_synth(&mut mat, &(section_size as usize), &false);
+    // let mut mat_t = mat.t().to_owned();
+    let upper_cnots = lower_cnot_synth(&mut mat, &(section_size as usize), &true);
+
+    // iterator over the gates
+    let instructions = upper_cnots
+        .iter()
+        .map(|(i, j)| (*j, *i))
+        .chain(lower_cnots.into_iter().rev())
+        .map(|(ctrl, target)| {
+            (
+                StandardGate::CXGate,
+                smallvec![],
+                smallvec![Qubit(ctrl as u32), Qubit(target as u32)],
+            )
+        });
+
+    Ok(
+        CircuitData::from_standard_gates(py, num_qubits as u32, instructions, Param::Float(0.0))
+            .expect("Something went sideways in Qiskit's Python realm!"),
```

Comment:
is it a standard Qiskit error?

Reply:
Yeah, from what I understand this must be handled every time we use the `from_standard_gates`, see e.g. https://github.com/mtreinish/qiskit-core/commit/4edf3fbd3d6f763053d67cc6dadec83f5174228f#diff-61c072175672ef132e5e682a51401abd08288938d864b2090f4dc446e772837eR855-R868 🙂 (Also the compiler is complaining otherwise)